frame/staking/src/lib.rs

```diff
@@ -524,14 +524,59 @@ impl<AccountId, Balance: HasCompact + Copy + Saturating + AtLeast32BitUnsigned +
 impl<AccountId, Balance> StakingLedger<AccountId, Balance>
 where
 	Balance: AtLeast32BitUnsigned + Saturating + Copy,
+	AccountId: Clone,
 {
-	/// Slash the validator for a given amount of balance. This can grow the value
-	/// of the slash in the case that the validator has less than `minimum_balance`
+	/// Slash the staker for a given amount of balance. This can grow the value
+	/// of the slash in the case that the staker has less than `minimum_balance`
 	/// active funds. Returns the amount of funds actually slashed.
 	///
 	/// Slashes from `active` funds first, and then `unlocking`, starting with the
 	/// chunks that are closest to unlocking.
-	fn slash(&mut self, mut value: Balance, minimum_balance: Balance) -> Balance {
+	fn slash<P: sp_staking::PoolsInterface<AccountId = AccountId, Balance = Balance>>(
```

Comment:
still considering doing this in another way.

Reply:
Nonetheless, you can just make `StakingLedger` be generic over `T: Config` to

1. have access to `BalanceOf<T>` and `T::AccountId`
2. have access to `T::PoolsInterface`

Reply:
Okay here's an idea: 

```
in staking config: 
type SlashingStrategy: Get<SlashingStrategy>;

/// Configuration of how the [`StakingLedger`] should be slashed.
enum SlashingStrategy {
  /// First slash the active bond, and then start with the potentially unlocking chunks. 
  ActiveFirst,
  /// Split the slash equally between the active and all the unlocking chunks. 
  SplitEqual
} 
```

Reply:
This kinda makes sense to me, because the two mentioned strategies are not totally random, and I can see some sense in them.

Other alternative is to move the entire slashing logic outside of the pallet and to the runtime. You basically add a `fn on_slash()` to your trait and let anyone implement it. the pallet should expose two public functions like `do_slash_active_first()` and `do_slash_all_chunks_equal()`. In a runtime without pools you implement this function with the former, in a runtime without slashing the latter. 

I kinda prefer the former for now, as it feels a bit strange to extract the entire slashing logic out of staking, but I find both of them better than the current way.

Reply:
I am not exactly sure how the slashing strategy would play in - don't we always want to maintain slash active first for anything that is not a pool? And then only use slash all chunks equal if we are slashing a pool?

Reply:
right, my suggestions where all in the direction of only supporting one, which renders them a bit different they your thinking. Just thinking out loud: what's wrong with just slashing everyone the same was as pools?

Reply:
My first reaction is we want to slash active first so their vote weight is immediatley decreased. Tbh I don't know all the rational but I don't think a marginally more complex implementation should dictate how we slash pools vs normal accounts.

Reply:
https://github.com/paritytech/substrate/pull/10982